livekit-agents/livekit/agents/codecs/mp3.py

```diff
@@ -0,0 +1,82 @@
+import queue
+import logging
+import ctypes
+
+class Mp3StreamDecoder:
+    def __init__(self):
+        try:
+            globals()["av"] = import_module("av")
+        except ImportError:
+            raise ImportError("You haven't included the decoder_utils optional dependencies. Please install the decoder_utils extra by running `pip install livekit-agents[decoder_utils]`")
+        self._input_queue = queue.Queue()
+        self._packet_queue = queue.Queue()
+        self._output_queue = queue.Queue()
+        self._closed = False
+        self._codec = av.CodecContext.create('mp3', 'r') # noqa
+        self._packet_parse_thread = threading.Thread(target=self._packet_parse_thread)
+        self._packet_parse_thread.start()
+        self._decode_thread = threading.Thread(target=self._decode_thread)
+        self._decode_thread.start()
+
+    def flush(self):
+        self._input_queue.put(None, block=False)
+
+    def close(self):
+        self._closed = True
+        self._output_queue.put(None, block=False)
+
+    def push_chunk(self, chunk: bytes):
+        if self._closed:
+            raise ValueError("Cannot push chunk to closed decoder")
+        self._input_queue.put(chunk, block=False)
+
+    def _packet_parse_thread(self):
+        while True:
+            data = self._input_queue.get()
+            if data is None:
+                self._packet_queue.put(None, block=False)
+                continue
+
+            packets = self._codec.parse(data)
+            for packet in packets:
+                self._packet_queue.put(packet, block=False)
+
+    def _decode_thread(self):
+        while True:
+            packet = self._packet_queue.get()
+            if packet is None and self._closed:
+                self._packet_queue.put(None, block=False)
+                break
+            try:
+                decoded = self._codec.decode(packet)
+                # flush case
+                if packet is None:
+                    print("creating new codec")
+                    self._codec = av.CodecContext.create('mp3', 'r') # noqa
+                    continue
+                print(decoded)
+                plane = decoded[0].planes[0]
+                ptr = plane.buffer_ptr
+                size = plane.buffer_size
+                byte_array_pointer = ctypes.cast(ptr, ctypes.POINTER(ctypes.c_char * size))
+                bytes_object = bytes(byte_array_pointer.contents)
+                self._output_queue.put(bytes_object, block=False)
+            except Exception as e:
+                logging.exception("Error decoding mp3 chunk", e)
+                continue
+
+    def __aiter__(self):
+        return self
+
+    async def __anext__(self):
+        while True:
+            try:
+                packet = self._output_queue.get_nowait()
+                if packet is None:
+                    raise StopAsyncIteration
+                return packet
+            except queue.Empty:
+                await asyncio.sleep(0.01)
```

Comment:
Not fan of waiting inside a loop, how slow is av.CodecContext if we want to use it inside the main thread?

Is there a reason to not use asyncio.to_thread?

Reply:
No I mean, can use use asyncio.to_thread to do the decode?

Reply:
Oh hrmm wdym?

So the input here is intended to be chunks of bytes that don't necessarily correspond to mp3 chunks. So we push them in a queue for a thread to consume them and create mp3 packets. Those mp3 packets then go into another queue (because they must stay in order) for decoding.

Both parsing and decoding are blocking and too slow for something you'd do on the same thread as the asyncio event loop.

Reply:
So I was thinking about creating a new task, but using to_thread inside it when starting to decode smth

Reply:
Ahh I see, made that adjustment. I also got rid of flush and made this class ephemeral. As I was using it, it was the more common use case (otherwise we'd have to introduce "started" and "finished" events)